File: PT1/receptor/src/correction/hamming.rs
#[derive(Debug)]
pub struct HammingStreamResult {
    pub data_bits: String,                    // concatenación de todos los bloques (solo datos)
    pub corrected_positions: Vec<(usize, usize)>, // (índice de bloque, posición corregida 1-based dentro del bloque)
}

// Devuelve true si x es potencia de dos (1, 2, 4, 8, ...)
fn is_power_of_two(x: usize) -> bool {
    x != 0 && (x & (x - 1)) == 0
}

// Calcula r tal que 2^r >= n + 1 (n = longitud del código).
// Este r corresponde al número de bits de paridad de un código de Hamming SEC (sin paridad global).
fn parity_bits_count(n: usize) -> usize {
    let mut r = 0usize;
    while (1usize << r) < (n + 1) {
        r += 1;
    }
    r
}

// Decodifica un bloque Hamming de longitud n (n = m + r).
// Retorna (datos, posición corregida) o error si el bloque es inconsistente.
fn decode_block(block: &[u8]) -> Result<(Vec<u8>, Option<usize>), String> {
    let n = block.len();
    if n < 3 {
        return Err("n demasiado pequeño".into());
    }
    let r = parity_bits_count(n);

    // Calcular síndrome
    // NOTA: El síndrome en códigos de Hamming es una secuencia de bits que se utiliza
    // para identificar y corregir errores en la transmisión de datos. La interpretación
    //  del síndrome revela la posición del bit erróneo, permitiendo su corrección. 
    let mut syndrome: usize = 0;
    for i in 0..r {
        // posición de paridad p = 2^i (1-based)
        let p = 1usize << i;
        let mut parity = 0u8;
        for pos in 1..=n {
            if (pos & p) != 0 {
                parity ^= block[pos - 1];
            }
        }
        if parity == 1 {
            syndrome |= p;
        }
    }

    // Si síndrome != 0, corregir ese bit (si está dentro de rango)
    let mut corrected_pos: Option<usize> = None;
    let mut corrected_block = block.to_vec();
    if syndrome != 0 {
        if syndrome >= 1 && syndrome <= n {
            corrected_block[syndrome - 1] ^= 1;
            corrected_pos = Some(syndrome);
        } else {
            return Err(format!("Síndrome {} fuera de rango para n={}", syndrome, n));
        }
    }

    // Extraer solo los bits de datos (omitir posiciones potencia de dos)
    let mut data = Vec::new();
    for pos in 1..=n {
        if !is_power_of_two(pos) {
            data.push(corrected_block[pos - 1]);
        }
    }

    Ok((data, corrected_pos))
}

// Decodifica una secuencia concatenada de bloques Hamming, cada uno de longitud n.
// Retorna todos los datos concatenados y las posiciones corregidas por bloque.
pub fn decode_stream(bits_str: &str, n: usize) -> Result<HammingStreamResult, String> {
    if !bits_str.chars().all(|c| c == '0' || c == '1') {
        return Err("Solo se aceptan '0' y '1'".to_string());
    }
    let bits: Vec<u8> = bits_str.chars().map(|c| if c == '1' {1} else {0}).collect();
    if bits.len() % n != 0 {
        return Err(format!("La longitud de la trama ({}) no es múltiplo de n={}.", bits.len(), n));
    }
    let num_blocks = bits.len() / n;

    let mut all_data = Vec::<u8>::new();
    let mut corrected_positions = Vec::<(usize, usize)>::new();

    for b in 0..num_blocks {
        let start = b * n;
        let end = start + n;
        let block = &bits[start..end];
        match decode_block(block) {
            Ok((data, corrected)) => {
                if let Some(pos) = corrected {
                    corrected_positions.push((b, pos));
                }
                all_data.extend_from_slice(&data);
            }
            Err(e) => {
                return Err(format!("Bloque {} inválido: {}", b + 1, e));
            }
        }
    }

    let data_bits: String = all_data.into_iter().map(|v| if v == 1 {'1'} else {'0'}).collect();
    Ok(HammingStreamResult { data_bits, corrected_positions })
}

File: PT1/emisor/src/correction/Hamming.java

package correction; 

public class Hamming {
    public static class EncodeResult {
        public final String encodedBits;
        public final int paddingZeros;
        public EncodeResult(String encodedBits, int paddingZeros) {
            this.encodedBits = encodedBits;
            this.paddingZeros = paddingZeros;
        }
    }
    private static boolean isPowerOfTwo(int x) {
        return x != 0 && (x & (x - 1)) == 0;
    }
    private static int parityBitsCount(int n) {
        int r = 0;
        while ((1 << r) < (n + 1)) r++;
        return r;
    }
    private static int[] parseBits(String s) {
        int[] out = new int[s.length()];
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c != '0' && c != '1') throw new IllegalArgumentException("Solo se aceptan '0' y '1'");
            out[i] = (c == '1') ? 1 : 0;
        }
        return out;
    }
    private static String bitsToString(int[] bits) {
        StringBuilder sb = new StringBuilder(bits.length);
        for (int b : bits) sb.append(b == 1 ? '1' : '0');
        return sb.toString();
    }
    public static EncodeResult encodeStream(String dataBits, int n) {
        int[] data = parseBits(dataBits);
        int r = parityBitsCount(n);
        int m = n - r;
        if (m <= 0) throw new IllegalArgumentException("n demasiado pequeño para código Hamming");

        int padding = (m - (data.length % m)) % m;
        int totalData = data.length + padding;

        int blocks = totalData / m;
        int[] dataPadded = new int[totalData];
        System.arraycopy(data, 0, dataPadded, 0, data.length);

        StringBuilder out = new StringBuilder(blocks * n);
        int idx = 0; 

        for (int b = 0; b < blocks; b++) {
            int[] codeword = new int[n];

            int dp = 0;
            for (int pos = 1; pos <= n; pos++) {
                if (!isPowerOfTwo(pos)) {
                    codeword[pos - 1] = dataPadded[idx + dp];
                    dp++;
                    if (dp == m) break;
                }
            }

            for (int i = 0; i < r; i++) {
                int p = 1 << i; 
                int parity = 0;
                for (int pos = 1; pos <= n; pos++) {
                    if ((pos & p) != 0) {
                        parity ^= codeword[pos - 1];
                    }
                }
                codeword[p - 1] = parity;
            }

            out.append(bitsToString(codeword));
            idx += m;
        }

        return new EncodeResult(out.toString(), padding);
    }
}
